File: C_Bakry_and_Partitioning.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <functional>

using namespace std;

void solve() {
    int t;
    cin >> t;

    while (t--) {
        int n, k;
        cin >> n >> k;
        vector<int> array(n);
        for (int i = 0; i < n; ++i) {
            cin >> array[i];
        }

        unordered_map<int, vector<int>> graph;
        for (int i = 0; i < n - 1; ++i) {
            int u, v;
            cin >> u >> v;
            graph[u].push_back(v);
            graph[v].push_back(u);
        }

        int target = 0;
        for (int i = 0; i < n; ++i) {
            target ^= array[i];
        }

        unordered_set<int> visited;
        int cuts = 0;

        function<int(int)> dfs = [&](int node) {
            if (visited.count(node)) return 0;
            visited.insert(node);

            int curr = array[node - 1];
            for (int neighbor : graph[node]) {
                curr ^= dfs(neighbor);
            }

            if (curr == target) {
                cuts++;
                return 0;
            }
            cout<<cuts<<endl;
            return curr; 
        };

        cout<<cuts<< endl;
        if (target == 0 || (dfs(1) == 0 && cuts > 1 && cuts < k )) {
            cout << "YES" << endl;
        } else {
            cout << "NO" << endl;
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    solve();

    return 0;
}

File: C_Bakry_and_Partitioning.py
from collections import defaultdict
t = int(input())
for _ in range(t):
    n,k = map(int,input().split())
    array = list(map(int,input().split()))
    graph = defaultdict(list)

    for _ in range(n - 1):
        u,v = map(int,input().split())
        graph[u].append(v)
        graph[v].append(u)


    print(graph)
    '''
        1 

        2  4

    3        5
        2 2 n, k
        1 3 array
        1 2   1 --> 2  target == 2
    '''
    target = 0
    for i in range(n):
        target ^= array[i]

    print(target)
    visited = set()
    cuts = 0
    def dfs(node):
        stack = [node]
        global cuts
        curr = 0
        while stack:
            node = stack.pop()
            if node in visited:
                continue
            
            visited.add(node)
            
            print(curr)
            for nei in graph[node]:
                stack.append(nei)

            curr ^= array[node - 1]
            print(curr,'before',node)
            if curr == target:
                print(curr,'about',target)
                cuts += 1
                curr = 0

        
        print(curr,target,cuts)
        return curr

    if not target or (not dfs(1) and cuts > 1):
        print('YES')
    else:
        print('NO')
    